app/resources/api/v2/qc_file_resource.rb

```diff
@@ -0,0 +1,126 @@
+    # or look at the [JSONAPI::Resources](http://jsonapi-resources.com/) package for Sequencescape's implementation
+    # of the JSON:API standard.
+    class QcFileResource < BaseResource
+      ###
+      # Attributes
+      ###
+
+      # @!attribute [r] content_type
+      #   @return [String] The content type, or MIME type, of the QC file.
+      attribute :content_type, readonly: true
+
+      # @!attribute [rw] contents
+      #   @return [String] The String contents of the QC file.
+      #      This is usually the CSV data for the QC file.
+      #      This can only be written once on creation.
+      attribute :contents, write_once: true
+
+      def contents
+        # The contents comes from the uploaded_data managed by CarrierWave.
+        @model.current_data
+      end
+
+      def contents=(value)
+        # Do not update the model.
+        # File contents is set via the uploaded_data hash supplied during QcFile creation.
+      end
+
+      # @!attribute [r] created_at
+      #   @return [DateTime] The date and time the QC file was created.
+      attribute :created_at, readonly: true
+
+      # @!attribute [rw] filename
+      #   @return [String] The filename of the QC file.
+      #      This can only be written once on creation.
+      attribute :filename, write_once: true
+
+      def filename=(value)
+        # Do not update the model.
+        # Filename is set via the uploaded_data hash supplied during QcFile creation.
+      end
+
+      # @!attribute [r] size
+      #   @return [Integer] The size of the QC file in bytes.
+      attribute :size, readonly: true
+
+      # @!attribute [r] uuid
+      #   @return [String] The UUID of the bulk transfers operation.
+      attribute :uuid, readonly: true
+
+      ###
+      # Relationships
+      ###
+
+      # @!attribute [rw] labware
+      #   @return [LabwareResource] The Labware which this QcFile belongs to.
+      has_one :labware, relation_name: :asset, foreign_key: :asset_id, write_once: true
+
+      ###
+      # Filters
+      ###
+
+      # @!method filter_uuid
+      #   Filter the QcFile resources by UUID.
+      #   @example GET request with UUID filter
+      #     GET /api/v2/qc_files?filter[uuid]=12345678-1234-1234-1234-123456789012
+      filter :uuid, apply: ->(records, value, _options) { records.with_uuid(value) }
+
+      ###
+      # Create method
+      ###
+
+      # @!method create_with_tempfile
+      #   Create a new QcFile resource with the uploaded data from a temporary file. This is called by the controller
+      #   when a create request for a QcFile is made. It ensures the contents of the file have been written to a
+      #   new TempFile instance.
+      # @param context [Hash] The context for the request.
+      # @param tempfile [Tempfile] A temporary file containing the uploaded data.
+      # @param filename [String] The filename for the uploaded data.
+      # @return [QcFileResource] The new QcFile resource.
```

Comment:
```suggestion
      # @return [QcFile] The new QcFile resource.
```
Is it returning a QcFile resource, as apposed to a QcFileResource resource?

Reply:
It is actually a `QcFileResource`.  I've found that confusing about these resources as well.  This is a class method which is creating an instance of the class.  The first `new` below creates the resource instance and passes it a `QcFile.new` and the `context` as parameters.  The first is then assigned to `@model` for use by instance methods that read or modify the underlying model object.